sympy/core/operations.py

```diff
@@ -504,3 +505,129 @@ def args(self):
+    ``_op_priority`` the priority relation of it and other classes should be registered in
+    this dispatcher by ``register_priority`` method.
+
+    Examples
+    ========
+
+    >>> from sympy import Add, Expr, Symbol
+    >>> from sympy.core.operations import AssocOpDispatcher
+    >>> newadd = AssocOpDispatcher("newadd", Add)
+
+    >>> class NewExpr(Expr):
+    ...     @classmethod
+    ...     def _newadd_handler(cls, *args, **kwargs):
+    ...         return NewAdd(*args, **kwargs)
+    >>> class NewAdd(NewExpr, Add):
+    ...     pass
+
+    >>> @newadd.register_priority(Expr, NewExpr)
+    ... @newadd.register_priority(NewExpr, Expr)
+    ... @newadd.register_priority(NewExpr, NewExpr)
+    ... def _(_1, _2):
+    ...     if issubclass(_1, NewExpr):
+    ...         return _1
+    ...     else:
+    ...         return _2
+
+    >>> a, b = Symbol('a'), NewExpr()
+    >>> newadd(a, a) == Add(a, a)
+    True
+    >>> newadd(a, b) == NewAdd(a, b)
+    True
+
+    See Also
+    ========
+
+    core.compatibility.OpDispatcher
+
+    """
+    def __init__(self, name, fallback, doc=None):
+        self.name = name
+        self.fallback = fallback
+        self.doc = doc
+        self.handlermethod = "_%s_handler" % name
+        self.dispatcher = Dispatcher(name)
+
+    def __repr__(self):
+        return "<dispatched %s>" % self.name
+
+    def register_priority(self, cls1, cls2):
+        """
+        Register the priority between two types. Registered function must
+        return the type which has the handler that is to be used.
+        """
+        def _(func):
+            self.add_priority((cls1, cls2), func)
+            return func
+        return _
+
+    def add_priority(self, signature, func, on_ambiguity=(lambda *args, **kwargs: None)):
+        # This suppresses unnecessary ambiguity warning, since not all sympy objects
+        # need to operate with each other.
+        return self.dispatcher.add(signature, func, on_ambiguity)
+
+    def __call__(self, *args, **kwargs):
+        types = tuple([type(arg) for arg in args])
+        return self.dispatch(types)(*args, **kwargs)
+
+    def dispatch(self, types):
+        """
+        Select the argument with highest priority, and return its handler.
+        """
+        args_data = []
+        for typ in types:
+            handler = getattr(typ, self.handlermethod, None)
+            if handler is None:
+                continue
+            op_priority = getattr(typ, '_op_priority', 0)
+            args_data.append((typ, op_priority, handler))
+
+        # No handler detected: use fallback
+        if len(args_data) == 0:
+            return self.fallback
+
+        # Sieve with _op_priority first
+        max_priority = max(op_priority for _, op_priority, _ in args_data)
+        sieved_data = [tup for tup in args_data if tup[1]==max_priority]
+
+        # Recursively select with registered binary priority
```

Comment:
This priority is a bit weird. It means that `mul(a,b,c)`, where a and c have tied priority, dispatches to `mul(a, c)` - even though a and c are not adjacent, and there's no reason to assume commutativity would allow them to be.

It seems to me that you should dispatch on all pairs, but use `_op_priority` to break the dispatcher ambiguities mentioned above.

Reply:
`AssocOpDispatcher`'s logic is different from `Dispatcher`, which directly returns the result. `mul(a,b,c)` does not store the multiplication logic in itself. Rather, its dispatcher selects the argument that presides the operation and delegates everything to it.

Suppose you want to multiply `a`(`vector.Vector`), `b`(`Integer`), and `c`(`TensExpr`). When you run `mul(a,b,c)`, this is what happens:

1. It compares `a._op_priority` (=12),  `b._op_priority` (=10), `c._op_priority` (=12). Since the max value is 12, `b` drops out while `a` and `c`  are in tie.

2. To break tie, `mul` checks if `Vector` and `TensExpr` are registered in its dispatcher. If they are meant to be multiplied, its dispatcher selects one from these two.

3. Let's suppose the dispatcher returned `Vector`. Then, `a`, `b`, `c` are passed to `Vector._mul_handler`. Everything after this is decided by it.

Therefore, whether elements are adjacent or not does not matter in this logic. Also, this is the reason why you don't need to register every combination. If you run `mul(b, c)`, `TensExpr._mul_handler` is called without referring to `mul`'s dispatcher because `c` is selected by `_op_priority`.

Reply:
> Also, this is the reason why you don't need to register every combination

My comment in the other thread is specifically about `OpDispatcher`, which I think ignores priority?

Reply:
Oh, sorry! I confused.